### PR TITLE
Add docker-server-less image name -> digest code

### PR DESCRIPTION
**Pull request by MrCreosote**
Uses the crane CLI tool, but abstracts it away from the API so we can switch it to something else (e.g. a python lib) later if possible without changing anything other than initialization code

### CODE REVIEW

File: cdmtaskservice/dockerimages.py
Comment (Tianhao-Gu) [inline cdmtaskservice/dockerimages.py:9]: can we use docker client? It will pull image but we don't need to install `crane`, validate image name/hostname, etc. 
  
```
In [5]: import docker
   ...: 
   ...: client = docker.from_env()

In [6]: image = client.images.pull("kbase/workspace_del
   ...: uxe:0.10.2-hotfix")

In [7]: image.attrs['RepoDigests'][0].split('@')[1]
Out[7]: 'sha256:285b1229730192eac5b502c44bfffff98e9840057821d1a26bed47a05bc44874'
```

File: cdmtaskservice/dockerimages.py
Comment (Tianhao-Gu) [inline cdmtaskservice/dockerimages.py:34]: Any executable (e.g., `git`, `docker`, etc.) that supports a valid `version` command can pass this validation.

File: cdmtaskservice/dockerimages.py
Comment (Tianhao-Gu) [inline cdmtaskservice/dockerimages.py:106]: Why won't images like `python:3.9-slim` be considered a valid image name for us?

File: test/dockerimages_test.py
Comment (Tianhao-Gu) [inline test/dockerimages_test.py:2]: The initial test run took over a minute to complete. After first run, it's pretty fast (4 seconds). `crane` doesn't pull anything at all locally? If the image is pulled locally, using the Docker client with just a few lines of code seems more reasonable.

```
collected 4 items                                     

test/dockerimages_test.py ....                  [100%]

============ 4 passed in 64.85s (0:01:04) =============
```

File: test.cfg.example
Comment (Tianhao-Gu) [inline test.cfg.example:6]: The online installation guide doesn’t provide any details on how to obtain the local path of the executable. Can you add instructions or an example here?

File: cdmtaskservice/dockerimages.py
Comment (MrCreosote) [reply cdmtaskservice/dockerimages.py:2]: It's to prevent arbitrary code execution on NERSC:

1. Create an image with tag & push to a repository
2. Register in the service after getting KBase staff blessing for the image
3. Push an entirely different image to that repository

File: cdmtaskservice/dockerimages.py
Comment (MrCreosote) [reply cdmtaskservice/dockerimages.py:9]: That's the first thing I tried but it won't work without a docker service, and the cdm task service won't have access to one.

File: cdmtaskservice/dockerimages.py
Comment (MrCreosote) [reply cdmtaskservice/dockerimages.py:34]: Yeah, this is more about making sure it can run and is there

File: cdmtaskservice/dockerimages.py
Comment (MrCreosote) [reply cdmtaskservice/dockerimages.py:106]: Hmm, good catch, although I doubt we'll ever use an image like that to run a cdm job, so maybe it's ok. What do you think?

File: test/dockerimages_test.py
Comment (MrCreosote) [reply test/dockerimages_test.py:2]: That's weird - maybe there was a network blip or something there? It always takes ~4.5 secs for me.

File: test.cfg.example
Comment (MrCreosote) [reply test.cfg.example:6]: It's just a tarball. Do you think we need to add instructions for that?

File: test/dockerimages_test.py
Comment (MrCreosote) [reply test/dockerimages_test.py:2]: > crane doesn't pull anything at all locally? 

No, it just queries the repository directly

Comment (MrCreosote) [reply test/dockerimages_test.py:2]: Looks like it takes <4s on GHA as well; I'm guessing it's either a blip or a mac thing?

https://github.com/kbase/cdm-task-service/actions/runs/10335338825/job/28609757953#step:5:25

File: cdmtaskservice/dockerimages.py
Comment (Tianhao-Gu) [reply cdmtaskservice/dockerimages.py:106]: I think that's okay. Maybe just add a note so that we know our limitations.

File: test/dockerimages_test.py
Comment (Tianhao-Gu) [reply test/dockerimages_test.py:2]: hmm yea I saw the GHA build.

File: test.cfg.example
Comment (Tianhao-Gu) [reply test.cfg.example:6]: I see. You downloaded the tar file. Hmmm. Maybe just a example string? I did brew install and `which crane` the get the path.

File: cdmtaskservice/dockerimages.py
Comment (Tianhao-Gu) [reply cdmtaskservice/dockerimages.py:34]: 👍 I wasn't able to resolve comments. Maybe I am not added to the repo yet.

File: cdmtaskservice/dockerimages.py
Comment (Tianhao-Gu) [reply cdmtaskservice/dockerimages.py:9]: 👍

File: cdmtaskservice/dockerimages.py
Comment (Tianhao-Gu) [reply cdmtaskservice/dockerimages.py:2]: But I guess an `image name` + `tag` won't work?

Comment (MrCreosote) [reply cdmtaskservice/dockerimages.py:2]: No, you can always push a different image to a name + tag combination. For example, you could tag & release something in GHA which would push the image, and then delete the release & tag and release and tag a completely different commit & push

File: cdmtaskservice/dockerimages.py
Comment (MrCreosote) [reply cdmtaskservice/dockerimages.py:34]: fixed

File: test.cfg.example
Comment (MrCreosote) [reply test.cfg.example:6]: > Maybe just a example string?

Not sure what you mean here?

File: cdmtaskservice/dockerimages.py
Comment (MrCreosote) [reply cdmtaskservice/dockerimages.py:106]: What's interesting is if you have something like `myrepo.io/myimage:tag` how do you know whether `myrepo.io` is a namespace or a tag?

File: cdmtaskservice/dockerimages.py
Comment (Tianhao-Gu) [reply cdmtaskservice/dockerimages.py:2]: yea. I know that. But in practice, we use `image name` + `tag` most of the time. I just don't see the benefits we gain justify the ongoing maintenance required for the hash. I am fine with it if you think it's the way to go. Hopefully we are not over-building it.

File: test.cfg.example
Comment (Tianhao-Gu) [reply test.cfg.example:6]: something like, `e.g. test.crane.exe=/opt/homebrew/bin/crane`

Comment (MrCreosote) [reply test.cfg.example:6]: Added some help text

File: cdmtaskservice/dockerimages.py
Comment (MrCreosote) [reply cdmtaskservice/dockerimages.py:109]: Updated the pydoc for the method

Comment (Tianhao-Gu) [reply cdmtaskservice/dockerimages.py:109]: 👍

File: test.cfg.example
Comment (Tianhao-Gu) [reply test.cfg.example:6]: 👍

File: cdmtaskservice/dockerimages.py
Comment (Tianhao-Gu) [reply cdmtaskservice/dockerimages.py:34]: 👍

**Review by Tianhao-Gu — APPROVED** [review]
LGTM.

File: cdmtaskservice/dockerimages.py
Comment (MrCreosote) [reply cdmtaskservice/dockerimages.py:2]: The benefit is not getting cut off by NERSC because someone was able to run malicious code on their compute because of our service

Comment (MrCreosote) [reply cdmtaskservice/dockerimages.py:2]: Also if we want longer lived SFAPI tokens we'll need to pass a security review; I doubt we'll pass if they discover that someone can run any code they want once they have one version of code approved

**Pull request merged**
